File: src/main/python/AoC2024_18.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 18
#

import sys
from collections.abc import Iterator

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.graph import bfs
from aoc.graph import flood_fill
from aoc.grid import Cell

Input = list[Cell]
Output1 = int
Output2 = str

SIZE = 71
TIME = 1024
START = Cell(0, 0)

TEST = """\
5,4
4,2
4,5
3,0
2,1
6,3
2,4
1,5
0,6
3,3
2,6
5,1
1,2
5,5
2,5
6,5
1,4
0,4
6,4
1,1
6,1
1,0
0,5
1,6
2,0
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [
            Cell(int(r), int(c))
            for c, r in (line.split(",") for line in input_data)
        ]

    def adjacent(
        self, cell: Cell, size: int, occupied: set[Cell]
    ) -> Iterator[Cell]:
        return (
            n
            for n in cell.get_capital_neighbours()
            if 0 <= n.row < size and 0 <= n.col < size and n not in occupied
        )

    def solve_1(self, cells: Input, size: int = SIZE, time: int = TIME) -> int:
        end = Cell(size - 1, size - 1)
        occupied = set(cells[:time])
        distance, _ = bfs(
            START,
            lambda cell: cell == end,
            lambda cell: self.adjacent(cell, size, occupied),
        )
        return distance

    def solve_2(self, cells: Input, size: int = SIZE, time: int = TIME) -> str:
        end = Cell(size - 1, size - 1)

        def free(time: int) -> set[Cell]:
            occupied = set(cells[:time])
            return flood_fill(
                START,
                lambda cell: self.adjacent(cell, size, occupied),
            )

        lo, hi = time, len(cells)
        while lo < hi:
            mid = (lo + hi) // 2
            if end in free(mid):
                lo = mid + 1
            else:
                hi = mid
        return f"{cells[lo - 1].col},{cells[lo - 1].row}"

    def part_1(self, cells: Input) -> Output1:
        return self.solve_1(cells)

    def part_2(self, cells: Input) -> Output2:
        return self.solve_2(cells)

    def samples(self) -> None:
        inputs = self.parse_input(TEST.splitlines())
        assert self.solve_1(inputs, 7, 12) == 22
        assert self.solve_2(inputs, 7, 12) == "6,1"


solution = Solution(2024, 18)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_18.java
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.graph.BFS;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

public final class AoC2024_18 extends SolutionBase<List<Cell>, Integer, String> {

    private static final int SIZE = 71;
    private static final int TIME = 1024;
    private static final Cell START = Cell.at(0, 0);

    private AoC2024_18(final boolean debug) {
        super(debug);
    }

    public static AoC2024_18 create() {
        return new AoC2024_18(false);
    }

    public static AoC2024_18 createDebug() {
        return new AoC2024_18(true);
    }

    @Override
    protected List<Cell> parseInput(final List<String> inputs) {
        return inputs.stream()
                .map(s -> StringOps.splitOnce(s, ","))
                .map(sp -> Cell.at(Integer.parseInt(sp.right()), Integer.parseInt(sp.left())))
                .toList();
    }

    private Stream<Cell> adjacent(final Cell cell, final int size, final Set<Cell> occupied) {
        return cell.capitalNeighbours()
                .filter(
                        n ->
                                0 <= n.getRow()
                                        && n.getRow() < size
                                        && 0 <= n.getCol()
                                        && n.getCol() < size
                                        && !occupied.contains(n));
    }

    private Integer solve1(final List<Cell> cells, final int size, final int time) {
        final Cell end = Cell.at(size - 1, size - 1);
        final Set<Cell> occupied = new HashSet<>(cells.subList(0, time));
        return BFS.execute(START, c -> c.equals(end), c -> adjacent(c, size, occupied));
    }

    private Set<Cell> free(final List<Cell> cells, final int size, final int time) {
        final Set<Cell> occupied = new HashSet<>(cells.subList(0, time));
        return BFS.floodFill(START, c -> adjacent(c, size, occupied));
    }

    private String solve2(final List<Cell> cells, final int size, final int time) {
        final Cell end = Cell.at(size - 1, size - 1);
        int lo = time;
        int hi = cells.size();
        while (lo < hi) {
            final int mid = (lo + hi) / 2;
            if (free(cells, size, mid).contains(end)) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        final Cell ans = cells.get(lo - 1);
        return String.format("%d,%d", ans.getCol(), ans.getRow());
    }

    @Override
    public Integer solvePart1(final List<Cell> input) {
        return solve1(input, SIZE, TIME);
    }

    @Override
    public String solvePart2(final List<Cell> input) {
        return solve2(input, SIZE, TIME);
    }

    @Override
    public void samples() {
        final AoC2024_18 test = AoC2024_18.createDebug();
        final List<Cell> input = test.parseInput(StringOps.splitLines(TEST));
        assert test.solve1(input, 7, 12) == 22;
        assert test.solve2(input, 7, 12).equals("6,1");
    }

    public static void main(final String[] args) throws Exception {
        AoC2024_18.create().run();
    }

    private static final String TEST =
            """
            5,4
            4,2
            4,5
            3,0
            2,1
            6,3
            2,4
            1,5
            0,6
            3,3
            2,6
            5,1
            1,2
            5,5
            2,5
            6,5
            1,4
            0,4
            6,4
            1,1
            6,1
            1,0
            0,5
            1,6
            2,0
            """;
}
